MegaApiClient/WebClient_HttpWebRequest.cs

```diff
@@ -1,89 +1,105 @@
-    {
-      HttpWebRequest request = this.CreateRequest(url);
-      request.ContentLength = dataStream.Length;
-      request.Method = "POST";
-      request.ContentType = contentType;
-
-      using (Stream requestStream = request.GetRequestStream())
-      {
-        dataStream.Position = 0;
-        dataStream.CopyTo(requestStream, this.BufferSize);
-      }
-
-      using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
-      {
-        using (Stream responseStream = response.GetResponseStream())
-        {
-          using (StreamReader streamReader = new StreamReader(responseStream, Encoding.UTF8))
-          {
-            return streamReader.ReadToEnd();
-          }
-        }
-      }
-    }
-
-    private HttpWebRequest CreateRequest(Uri url)
-    {
-      HttpWebRequest request = (HttpWebRequest)WebRequest.Create(url);
-      request.Timeout = this.responseTimeout;
-      request.UserAgent = this.userAgent;
-
-      return request;
-    }
-
-    private string GenerateUserAgent()
-    {
-      AssemblyName assemblyName = Assembly.GetExecutingAssembly().GetName();
-      return string.Format("{0} v{1}", assemblyName.Name, assemblyName.Version.ToString(2));
-    }
-  }
-}
+﻿#if NET40
+namespace CG.Web.MegaApiClient
+{
+  using System;
+  using System.IO;
+  using System.Net;
+  using System.Reflection;
+  using System.Text;
+  using System.Threading;
+
+  public class WebClient : IWebClient
+  {
+    private const int DefaultResponseTimeout = Timeout.Infinite;
+
+    private readonly int responseTimeout;
+    private readonly string userAgent;
+
+    public WebClient(int responseTimeout = DefaultResponseTimeout, string userAgent = null)
+    {
+      this.BufferSize = Options.DefaultBufferSize;
+      this.responseTimeout = responseTimeout;
+      this.userAgent = userAgent ?? this.GenerateUserAgent();
+    }
+
+    public int BufferSize { get; set; }
+
+    public string PostRequestJson(Uri url, string jsonData)
+    {
+      using (MemoryStream jsonStream = new MemoryStream(jsonData.ToBytes()))
+      {
+        return this.PostRequest(url, jsonStream, "application/json");
+      }
+    }
+
+    public string PostRequestRaw(Uri url, Stream dataStream)
+    {
+      return this.PostRequest(url, dataStream, "application/octet-stream");
+    }
+
+    public Stream GetRequestRaw(Uri url)
+    {
+      HttpWebRequest request = this.CreateRequest(url);
+      request.Method = "GET";
+
+      return request.GetResponse().GetResponseStream();
+    }
+
+    public long GetLength(Uri url)
+    {
+      HttpWebRequest request = this.CreateRequest(url);
+      request.Method = "HEAD";
+      return request.GetResponse().ContentLength;
+    }
+
+    public Stream GetRequestRawWithRange(Uri url, long startByte, long endByte)
+    {
+      HttpWebRequest request = this.CreateRequest(url);
+      request.Method = "GET";
+      request.AddRange(startByte, endByte - 1);
+
+      return request.GetResponse().GetResponseStream();
```

Comment:
It seems the Stream cannot be accessed properly from the SeekableStream class. I checked the differences with WebClient_HttpClient.cs and the HttpClient implementation returns a memory stream.
So could you replace this line with
```csharp
// Introduced an intermediate stream because the response stream
// is not seekable and has no valid Length property
MemoryStream memoryStream = new MemoryStream();

using (var netStream = request.GetResponse().GetResponseStream())
{
  netStream.CopyTo(memoryStream);
}

memoryStream.Position = 0;
return memoryStream;
```